pytorch_lightning/metrics/functional/classification.py

```diff
@@ -632,6 +638,83 @@ def _multiclass_auroc(pred, target, sample_weight, num_classes):
     return torch.mean(class_aurocs)
 
 
+def object_detection_mean_average_precision(
+        preds: torch.Tensor, target: torch.Tensor, iou_threshold: float, ap_calculation: str
+) -> torch.Tensor:
+    """
+    Compute mean average precision for object detection task. Calculates average precision
+    using AUC.
+
+    Args:
+        preds: an Nx7 batch of predictions with representation
```

Comment:
Don't we have different types of information here in the same `preds` tensor? Image index and class label are integers, confidence score / probability is a float, and the box coordinates may be integers or floats, depending on the data. Is it better to have separate tensors `label_preds`, `box_preds`, and `confidences`, so that they can have different data types?

Reply:
It's not immediately obvious to me why having separate tensors with different data types is of enough importance that it needs to be split into multiple arguments. I think having just `preds` and `target` is less confusing and more congruent with the other metrics in PyTorch Lightning. What do you think @Borda?

Reply:
I can see that in the test you use a float tensor, meaning that the image index and class label are also floats. In that case you do a lot of comparisons of equality with floats, like `preds[:, 1] == c`, `unique()` and `Counter()` called on floats, and the `targets_assigned` dictionary has float keys. That looks scary to me. If you want to keep the interface like it is, you could cast them to integers in the beginning of the function, like this:

    label_preds = preds[:, 1].to(dtype=torch.int64)

But then again, we would be concatenating a bunch of different stuff with different data types into a float tensor, just to call a function that will split them into separate variables and convert to correct data types, so I don't know if that would make a lot of sense either. :)

I didn't find existing code that would have exactly the same kind of situation. Maybe the closest thing I found is the [Faster R-CNN model](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/detection/faster_rcnn.py#L37) in Bolts that gets boxes in a float tensor and labels in an integer tensor. Anyway, I'm also interested to hear opinions from others.

Reply:
By the way, I would move this metric to a new file `detection.py`, as the other metrics in this file operate on classification predictions and targets, and the `iou()` and `average_precision()` in this file do something else than what they mean in object detection.